File: src/main/java/io/github/mg138/tsbook/items/ItemStats.java
package io.github.mg138.tsbook.items;

import io.github.mg138.tsbook.Book;
import io.github.mg138.tsbook.config.item.element.ItemSetting;
import io.github.mg138.tsbook.config.item.element.StatedItemSetting;
import io.github.mg138.tsbook.config.AbstractConfig;
import io.github.mg138.tsbook.items.data.stat.*;

public class ItemStats {
    private final ItemIdentification identification;
    private final AbstractConfig config;
    private final StatMap stats;

    public Double getStat(StatType type) {
        return getStats().get(type).getStat() * getIdentification().getStatPercentage(type);
    }

    public StatMap getStats() {
        return stats;
    }

    public ItemIdentification getIdentification() {
        return identification;
    }

    public String translate(StatType type, Stat stat) {
        String format = config.translate.translate("format." + type.toString());
        double percentage = identification.getStatPercentage(type);

        if (stat instanceof StatRange) {
            StatRange statRange = (StatRange) stat;
            return format
                    .replace("[min]", String.valueOf((int) (statRange.getMin() * percentage)))
                    .replace("[max]", String.valueOf((int) (statRange.getMax() * percentage)))
                    .replace("[percentage]", String.valueOf((int) (percentage * 100)) + '%');
        } else {
            return format
                    .replace("[stat]", String.valueOf((int) (stat.getStat() * percentage)))
                    .replace("[percentage]", String.valueOf((int) (percentage * 100)) + '%');
        }
    }

    public ItemStats(ItemIdentification identification, AbstractConfig config, StatMap stats) {
        this.identification = identification;
        this.config = config;
        this.stats = stats;
    }

    public static ItemStats create(ItemIdentification identification, AbstractConfig config, ItemSetting setting) {
        if (setting instanceof StatedItemSetting) {
            return new ItemStats(identification, config, ((StatedItemSetting) setting).STATS);
        } else return null;
    }

    public static ItemStats create(ItemIdentification identification, AbstractConfig config, String ID) {
        return create(
                identification,
                config,
                Book.Companion.getCfg().itemConfig.getAnyItemByID(ID)
        );
    }
}

File: src/main/java/io/github/mg138/tsbook/items/ItemStats.kt
package io.github.mg138.tsbook.items

import io.github.mg138.tsbook.setting.AbstractConfig
import io.github.mg138.tsbook.setting.item.ItemConfig
import io.github.mg138.tsbook.setting.item.element.ItemSetting
import io.github.mg138.tsbook.setting.item.element.StatedItemSetting
import io.github.mg138.tsbook.stat.*
import java.lang.IllegalArgumentException

class ItemStats(initStats: StatMap, val identification: ItemIdentification, private val config: AbstractConfig) {
    val stats = StatMap()

    init {
        initStats.forEach { (type, stat) ->
            val multiplier = identification[type]
            stats[type] = when {
                multiplier == 0.0F -> stat
                multiplier > 0.0F -> stat * multiplier
                else -> throw IllegalArgumentException("Identification must be >= 0! (is $multiplier, keyed to $type)")
            }
        }
    }

    operator fun get(type: StatType): Stat {
        return stats[type] ?: throw NullPointerException("No value keyed to $type.")
    }

    fun translate(type: StatType, stat: Stat): String {
        val format = config.translate.translate("format.$type")
        val percentage = identification[type]
        return when (val that = stat * percentage) {
            is StatRange -> {
                format
                    .replace("[min]", (that.min).toInt().toString())
                    .replace("[max]", (that.max).toInt().toString())
                    .replace("[percentage]", (percentage * 100).toInt().toString() + '%')
            }
            is StatSingle -> {
                format
                    .replace("[stat]", (that.stat).toInt().toString())
                    .replace("[percentage]", (percentage * 100).toInt().toString() + '%')
            }
            else -> throw IllegalArgumentException("Unsupported implementation of Stat interface. (Did someone forget to add it?)")
        }
    }

    companion object {
        fun create(setting: ItemSetting, identification: ItemIdentification?, config: AbstractConfig): ItemStats? {
            return if (setting is StatedItemSetting && identification != null) {
                ItemStats(setting.stats, identification, config)
            } else null
        }

        fun create(setting: ItemSetting, config: AbstractConfig, isRandom: Boolean): ItemStats? {
            return create(
                setting,
                ItemIdentification.create(setting, isRandom),
                config
            )
        }

        fun create(identification: ItemIdentification?, config: AbstractConfig, ID: String): ItemStats? {
            return create(
                ItemConfig.getAnyItemByID(ID),
                identification,
                config
            )
        }
    }
}